app/feature/feature-forever/src/main/kotlin/com/hedvig/android/feature/forever/ui/ReferralList.kt

```diff
@@ -0,0 +1,183 @@
+
+import androidx.compose.foundation.background
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.Row
+import androidx.compose.foundation.layout.Spacer
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.foundation.layout.padding
+import androidx.compose.foundation.layout.size
+import androidx.compose.foundation.layout.width
+import androidx.compose.foundation.layout.wrapContentSize
+import androidx.compose.foundation.shape.CircleShape
+import androidx.compose.material.LocalTextStyle
+import androidx.compose.material.Surface
+import androidx.compose.material.Text
+import androidx.compose.material3.Divider
+import androidx.compose.material3.LocalContentColor
+import androidx.compose.material3.MaterialTheme
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.CompositionLocalProvider
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.style.TextDecoration
+import androidx.compose.ui.unit.dp
+import com.hedvig.android.apollo.format
+import com.hedvig.android.core.designsystem.material3.typeElement
+import com.hedvig.android.core.designsystem.material3.warningElement
+import com.hedvig.android.core.designsystem.preview.HedvigPreview
+import com.hedvig.android.core.designsystem.theme.HedvigTheme
+import com.hedvig.android.core.ui.getLocale
+import com.hedvig.android.feature.forever.ForeverUiState
+import hedvig.resources.R
+import javax.money.CurrencyContext
+import javax.money.CurrencyUnit
+import org.javamoney.moneta.Money
+
+@Composable
+internal fun ReferralList(uiState: ForeverUiState) {
+  if (uiState.referrals.isEmpty()) {
+    return
+  }
+
+  Column(modifier = Modifier.padding(horizontal = 16.dp)) {
+    Text(
+      text = stringResource(id = R.string.FOREVER_REFERRAL_LIST_LABEL),
+      modifier = Modifier.padding(vertical = 16.dp),
+    )
+    Divider()
+    uiState.referrals.forEach { referral ->
+      ReferralRow(referral)
+    }
+    if (uiState.currentDiscountAmount != null && !uiState.currentDiscountAmount.isZero) {
+      Row(
+        modifier = Modifier
+          .padding(vertical = 16.dp)
+          .fillMaxWidth(),
+        horizontalArrangement = Arrangement.SpaceBetween,
+      ) {
+        Text(text = stringResource(id = R.string.referrals_your_other_discounts))
+        Text(uiState.currentDiscountAmount.format(getLocale()))
+      }
+      Divider()
+    }
+    Row(
+      modifier = Modifier
+        .padding(vertical = 16.dp)
+        .fillMaxWidth(),
+      horizontalArrangement = Arrangement.SpaceBetween,
+    ) {
+      Text(text = stringResource(id = R.string.FOREVER_TAB_TOTAL_DISCOUNT_LABEL))
+      Row {
+        CompositionLocalProvider(LocalContentColor provides MaterialTheme.colorScheme.onSurfaceVariant) {
```

Comment:
Not sure if this is the correct way of setting the text color? It should be gray, but im not getting that with this currently

Reply:
Ah yeah this is interesting.

So Text() has somewhat of an order in which it looks at to know what color it should have.

Text composable has a `color: Color = Color.Unspecified,` parameter, which you can pass a color directly in it.
Then, TextStyle also has a `color: Color = Color.Unspecified,` inside of it.
Finally, we got LocalContentColor which also is a color.

If you don't provide it anything `Text("someText")`, it will try and use the LocalTextStyle and see if it has a color inside there. If it does not, then it will look inside `LocalContentColor.current`.
Our styles [here](https://github.com/HedvigInsurance/android/blob/7e85f0db5781027e32700587c32c517454c95dac/app/core/core-design-system/src/main/kotlin/com/hedvig/android/core/designsystem/material3/HedvigTypography.kt#L23) don't provide any style-specific colors, so by default TextStyle will never have color in our app at least.

Then you can call your text like this
```
Text(
 "foo",
 style = LocalTextStyle.current.copy(color = someColor)
)
```
then since there is a color inside textStyle, it won't fall down to using LocalContentColor, and use the color inside the style directly.

Finally, if you call your text like this:
```
Text(
  "foo",
  color = someColor,
)
```
Then that takes precedence over everything,

To summarize this:
```
CompositionLocalProvider(LocalContentColor provides Color.Yellow) {
  Text(
    text = "foo",
    color = Color.Red,
    style = LocalTextStyle.current.copy(color = Color.Blue)
  )
}
```
Will *not* use the local content color, since a style color exists.
Will *not* use the style color, since a specific color for this particular text composable exists
*Will* use Color.Red since that's the specific color passed to this composable.

It's done this way so that if you don't have any styles, it at least falls back to LocalContentColor so that it's light on dark surfaces, and dark on light surfaces.
And so that you can define a style-wide color for the entire app if you wish using TextStyle.
And you can also pass specific ones for specific composables you wish.

tl;dr: For these situations you most likely just wanna do:
```
Text(
  text = "",
  color = MaterialTheme.colorScheme.onSurfaceVariant,
)
```

And in situations where you've got a whole section which should have these colors, just do:
```
CompositionLocalProvider(LocalContentColor provides MaterialTheme.colorScheme.onSurfaceVariant) {
  Column {
    Text("I am gray")
    Text("I am also gray")
    Text("I am also gray")
    Text("I am also gray")
  }
}
```

Reply:
Awesome, thanks 💯